File: 1840. Maximum Building Height.js
/**
 * @param {number} n
 * @param {number[][]} restrictions
 * @return {number}
 */
var maxBuilding = function (n, restrictions) {

    // Add building 1 and building n
    restrictions.push([1, 0]);
    restrictions.push([n, n - 1]);

    // Sort by building index
    restrictions.sort((a, b) => a[0] - b[0]);

    const m = restrictions.length;

    // Left to Right pass
    for (let i = 1; i < m; i++) {

        restrictions[i][1] = Math.min(
            restrictions[i][1],
            restrictions[i - 1][1] +
            (restrictions[i][0] - restrictions[i - 1][0])
        );
    }

    // Right to Left pass
    for (let i = m - 2; i >= 0; i--) {

        restrictions[i][1] = Math.min(
            restrictions[i][1],
            restrictions[i + 1][1] +
            (restrictions[i + 1][0] - restrictions[i][0])
        );
    }

    let answer = 0;

    // Find maximum possible peak
    for (let i = 1; i < m; i++) {

        const [x1, h1] = restrictions[i - 1];
        const [x2, h2] = restrictions[i];

        const dist = x2 - x1;

        const peak = Math.floor((h1 + h2 + dist) / 2);

        answer = Math.max(answer, peak);
    }

    return answer;
};

File: 1840. Maximum Building Height.py
class Solution:
    def maxBuilding(self, n: int, restrictions: list[list[int]]) -> int:

        # Add building 1 and building n as restrictions
        restrictions.append([1, 0])
        restrictions.append([n, n - 1])

        # Sort restrictions by building index
        restrictions.sort()

        m = len(restrictions)

        # Left to Right pass
        # Ensure current restriction is reachable
        for i in range(1, m):
            restrictions[i][1] = min(
                restrictions[i][1],
                restrictions[i - 1][1] +
                (restrictions[i][0] - restrictions[i - 1][0])
            )

        # Right to Left pass
        # Tighten restrictions from the right side
        for i in range(m - 2, -1, -1):
            restrictions[i][1] = min(
                restrictions[i][1],
                restrictions[i + 1][1] +
                (restrictions[i + 1][0] - restrictions[i][0])
            )

        answer = 0

        # Find maximum possible peak between adjacent restrictions
        for i in range(1, m):

            x1, h1 = restrictions[i - 1]
            x2, h2 = restrictions[i]

            dist = x2 - x1

            peak = (h1 + h2 + dist) // 2

            answer = max(answer, peak)

        return answer
